### PR TITLE
Consume tsc config at Signal Optimization Service 

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description

Update **Signal Optimization Service** to consume Traffic Signal Controller (TSC) configuration information from **TSC Service**.
## Related Issue
#219 

## Motivation and Context

**Signal Optimization Service** needs information about TSC configuration such as red clearance interval, yellow change and concurrent signal groups to create valid movement groups and valid desired phase plan durations.
## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: scheduling_service/src/scheduling_service.cpp [hunk 1]
```diff
@@ -173,15 +173,11 @@ namespace scheduling_service{
         {
             
             const std::string payload = consumer_worker->consume(1000);
-            try {
             if(payload.length() != 0 && vehicle_list_ptr)
             {                
 
                 vehicle_list_ptr->process_update(payload);
```
Comment (paulbourelly999) [inline scheduling_service/src/scheduling_service.cpp:179]: process_update from vehicle list does not throw any exception so try catch is not reachable

File: signal_opt_service/include/signal_opt_service.h [hunk 2]
```diff
@@ -39,14 +42,37 @@ namespace signal_opt_service
...
          * @brief Create threads and consume messages
          */
-        void start() const;
-
+        void start();
+        /**
+         * @brief Method to consume SPaT JSON from kafka consumer and update 
+         * spat pointer in signal_opt_message_worker. 
+         * 
+         */
```
Comment (paulbourelly999) [inline signal_opt_service/include/signal_opt_service.h:50]: Need to update javadoc comment for these methods.

File: signal_opt_service/src/signal_opt_messages_worker.cpp [hunk 3]
```diff
@@ -50,7 +71,11 @@ namespace signal_opt_service
                     if(invalid_signal_group_count == 0)
                     {
                         //Update intersection info 
-                        this->intersection_info_ptr = std::make_shared<OpenAPI::OAIIntersection_info>(int_info);
```
Comment (paulbourelly999) [inline signal_opt_service/src/signal_opt_messages_worker.cpp:53]: copy constructor seems to be delete for this class.

File: signal_opt_service/src/signal_opt_service.cpp [hunk 4]
```diff
@@ -55,50 +66,86 @@ namespace signal_opt_service
...
-        spat_t.join();
-        vsi_t.join();
+        std::thread spat_t(&signal_opt_service::consume_spat,this, this->_spat_consumer, this->spat_ptr);
+        std::thread vsi_t(&signal_opt_service::consume_vsi,  this, this->_vsi_consumer, this->vehicle_list_ptr);
+        std::thread tsc_config_t(&signal_opt_service::consume_tsc_config, this, this->_tsc_config_consumer, this->tsc_configuration_ptr);
+        // Get TSC Information then populate movement groups.
+        tsc_config_t.join();
+        populate_movement_groups(_movement_groups, tsc_configuration_ptr);
+        
+        spat_t.detach();
```
Comment (paulbourelly999) [inline signal_opt_service/src/signal_opt_service.cpp:78]: Threads need to be running in parallel so we can't use join.

File: signal_opt_service/src/signal_opt_messages_worker.cpp (hunk 3 above)
Comment (paulbourelly999) [reply signal_opt_service/src/signal_opt_messages_worker.cpp:53]: May need to use std::move for this. Will try and correct.

File: signal_opt_service/src/signal_opt_service.cpp [hunk 5]
```diff
@@ -7,6 +7,12 @@ namespace signal_opt_service
         try
         {
             _so_msgs_worker_ptr = std::make_shared<signal_opt_messages_worker>();
+            _movement_groups = std::make_shared<movement_groups>();
+            intersection_info_ptr = std::make_shared<OpenAPI::OAIIntersection_info>();
+            vehicle_list_ptr = std::make_shared<streets_vehicles::vehicle_list>();
+            vehicle_list_ptr->set_processor(std::make_shared<streets_vehicles::signalized_status_intent_processor>());
```
Comment (Saeid-Solei) [inline signal_opt_service/src/signal_opt_service.cpp:13]: I think you should read the exp_delta from manifest.json and set the processor timeout as well.

File: tsc_client_service/src/tsc_service.cpp [hunk 6]
```diff
@@ -206,13 +206,21 @@ namespace traffic_signal_controller_service {
     }
 
     void tsc_service::produce_tsc_config_json() {
-        // Publish tsc_config information 10 times
+        // Get configuration parameter value
         try {
-            while(tsc_config_state_ptr && tsc_config_producer_counter_ < 10)
+            tsc_config_send_attempts = streets_service::streets_configuration::get_int_config("tsc_config_send_attempts");
```
Comment (Saeid-Solei) [inline tsc_client_service/src/tsc_service.cpp:211]: Do we need to have a separate try\catch for just reading from manifest.json?

File: signal_opt_service/manifest.json [hunk 7]
```diff
@@ -43,6 +43,18 @@
...
+            "value": "tsc_config_state",
+            "description": "Kafka topic for CARMA Streets internal traffic signal controller phase configuration message.",
+            "type": "STRING"
+        },
+        {
+            "name": "tsc_config_group_id",
+            "value": "tsc_config_group",
+            "description": "Kafka consumer group for SO tsc_config_consumer.",
+            "type": "STRING"
         }
```
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply signal_opt_service/manifest.json:58]: Good catch

File: signal_opt_service/src/signal_opt_service.cpp (hunk 5 above)
Comment (paulbourelly999) [reply signal_opt_service/src/signal_opt_service.cpp:13]: good catch

File: tsc_client_service/src/tsc_service.cpp (hunk 6 above)
Comment (paulbourelly999) [reply tsc_client_service/src/tsc_service.cpp:211]: Yes, for error handling, having specific catch clauses for different exceptions helps our service respond better to errors.

File: scheduling_service/src/scheduling_service.cpp [hunk 8]
```diff
@@ -173,15 +173,11 @@ namespace scheduling_service{
...
     
-            }}
-            catch(const streets_vehicles::status_intent_processing_exception &e) {
```
Comment (dan-du-car) [inline scheduling_service/src/scheduling_service.cpp:183]: Why is this removed?

File: signal_opt_service/CMakeLists.txt [hunk 9]
```diff
@@ -74,6 +76,7 @@ target_link_libraries(${BINARY} PUBLIC
     spdlog::spdlog
     rapidjson
     gtest
+    gmock
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply signal_opt_service/CMakeLists.txt:79]: this -> gmock? Why do we need in the install process? Is the mock_kafka_consumer used for unit testing or?

File: signal_opt_service/include/movement_group.h [hunk 10]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+#include <list>
+
+namespace signal_opt_service
+{
+    struct movement_group
+    {
+        std::string name;
```
Comment (dan-du-car) [inline signal_opt_service/include/movement_group.h:8]: Please add comments? What is the name for this movement group?

File: signal_opt_service/include/movement_group.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+        std::pair<int,int> signal_groups;
```
Comment (dan-du-car) [inline signal_opt_service/include/movement_group.h:10]: Why signal groups is a pair of two int? not a list of int? Maybe add description like each movement group includes two signal group ids only?

File: signal_opt_service/include/movement_group.h [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    };
+
+    struct movement_groups {
+        std::list<movement_group> groups;
```
Comment (dan-du-car) [inline signal_opt_service/include/movement_group.h:14]: Please add comments about the groups? or maybe rename struct to movement_group_list which has  a member called movement_group?

File: signal_opt_service/src/signal_opt_service.cpp [hunk 13]
```diff
@@ -55,50 +68,102 @@ namespace signal_opt_service
...
 
-    void signal_opt_service::start() const
+    void signal_opt_service::start()
     {
-        std::thread spat_t(&signal_opt_service::consume_msg, this, std::ref(_spat_consumer), CONSUME_MSG_TYPE::SPAT);
-        std::thread vsi_t(&signal_opt_service::consume_msg, this, std::ref(_vsi_consumer), CONSUME_MSG_TYPE::VEHICLE_STATUS_INTENT);
+        std::thread tsc_config_t(&signal_opt_service::consume_tsc_config, this, this->_tsc_config_consumer, this->tsc_configuration_ptr);
+        // Get TSC Information then populate movement groups.
+        tsc_config_t.join();
+        populate_movement_groups(_movement_groups, tsc_configuration_ptr);
```
Comment (dan-du-car) [inline signal_opt_service/src/signal_opt_service.cpp:76]: What will happen if the tsc_config_t is failed ? will the populate_movement_groups not called?
Maybe add an inline comments about the join() before call the populate_movement_groups. If  _tsc_config_consumer receives more update the tsc_config after first update, the populate_movement_groups will not be called any more?

File: signal_opt_service/src/signal_opt_service.cpp [hunk 14]
```diff
@@ -55,50 +68,102 @@ namespace signal_opt_service
...
-                    }
-                    break;
-                case CONSUME_MSG_TYPE::VEHICLE_STATUS_INTENT:
-                    if (!_so_msgs_worker_ptr->add_update_vehicle(payload))
-                    {
-                        SPDLOG_CRITICAL("Error occurred when updating vehicle list.");
+    void signal_opt_service::consume_vsi(const std::shared_ptr<kafka_clients::kafka_consumer_worker> vsi_consumer,
+                                         const std::shared_ptr<streets_vehicles::vehicle_list> _vehicle_list_ptr) const{
+        while( vsi_consumer->is_running()) {
+            const std::string payload = vsi_consumer->consume(1000);
```
Comment (dan-du-car) [inline signal_opt_service/src/signal_opt_service.cpp:99]: Not sure why cannot just reuse the same function?

File: signal_opt_service/src/signal_opt_service.cpp [hunk 15]
```diff
@@ -107,14 +172,14 @@ namespace signal_opt_service
...
-            if (_so_msgs_worker_ptr && _so_msgs_worker_ptr->request_intersection_info())
+            if (_so_msgs_worker_ptr && _so_msgs_worker_ptr->request_intersection_info(intersection_info_ptr))
             {
                 return true;
             }
             sleep(sleep_secs);
             attempt_count++;
         }
-        // If failed to update the intersection information after certain numbers of attempts
+        // If failed to update the intersecstd::stringtion information after certain numbers of attempts
```
Comment (dan-du-car) [inline signal_opt_service/src/signal_opt_service.cpp:182]: intersecstd::stringtio?

File: signal_opt_service/test/test_signal_opt_messages_worker.cpp [hunk 16]
```diff
@@ -3,81 +3,139 @@
...
+    SPDLOG_INFO("Assert empty list!");
     auto cur_time = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count() + 1000;
+    // Valid payload
     std::string vsi_payload = "{ \"metadata\": { \"timestamp\": " + std::to_string(cur_time) + " }, \"payload\": { \"v_id\": \"DOT-507\", \"v_length\": 5, \"min_gap\": 10.0, \"react_t\": 1.5, \"max_accel\": 5.0, \"max_decel\": 5.0, \"cur_speed\": 5.0, \"cur_accel\": 0.0, \"cur_lane_id\": 7, \"cur_ds\": 7.0, \"direction\": \"straight\", \"entry_lane_id\": 7, \"link_lane_id\": 8, \"dest_lane_id\": 9, \"is_allowed\": false, \"depart_pos\": 1, \"est_paths\": [{ \"ts\": 1623677096200, \"id\": 7, \"ds\": 6.0 }, { \"ts\": 1623677096400, \"id\": 7, \"ds\": 5.0 }, { \"ts\": 1623677096600, \"id\": 7, \"ds\": 4.0 }, { \"ts\": 1623677096800, \"id\": 7, \"ds\": 3.0 }, { \"ts\": 1623677097000, \"id\": 7, \"ds\": 2.0 } ] } }";
-    so_msgs_worker_ptr->add_update_vehicle(vsi_payload);
-    ASSERT_EQ(1, so_msgs_worker_ptr->get_vehicle_list()->get_vehicles().size());
+    so_msgs_worker_ptr->update_vehicle_list(vsi_payload, vehicle_list_ptr);
+    SPDLOG_INFO("Populated vehicle list");
```
Comment (dan-du-car) [inline signal_opt_service/test/test_signal_opt_messages_worker.cpp:20]: maybe SPDLOG_DEBUG?

Comment (paulbourelly999) [reply signal_opt_service/test/test_signal_opt_messages_worker.cpp:20]: Cannot set loglevel for unit testing. Defaults to info.

File: signal_opt_service/CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply signal_opt_service/CMakeLists.txt:79]: Yes we need it in the install process. Unfortunately we haven't setup an cmake install target for kafka_clients library that includes its dependencies.  I have had trouble getting a cmake install target for kafka_clients because I could not figure out how to cmake install rdkafka to be able to find it with `find_package`. (https://github.com/edenhill/librdkafka/blob/master/packaging/cmake/README.md) The CMake install instructions included on their page do not seem to work for me right now.

File: scheduling_service/src/scheduling_service.cpp (hunk 8 above)
Comment (paulbourelly999) [reply scheduling_service/src/scheduling_service.cpp:183]: `vehicle_list::process_update` already catches this exception so this code is not reachable.

File: signal_opt_service/include/movement_group.h (hunk 11 above)
Comment (paulbourelly999) [reply signal_opt_service/include/movement_group.h:10]: Each movement group only includes two signal group ids. I can add a comment.

File: signal_opt_service/include/movement_group.h (hunk 12 above)
Comment (paulbourelly999) [reply signal_opt_service/include/movement_group.h:14]: Will do

File: signal_opt_service/src/signal_opt_service.cpp (hunk 13 above)
Comment (paulbourelly999) [reply signal_opt_service/src/signal_opt_service.cpp:76]: I will add more comments. Currently we are not supporting multiple updates to tsc_config. We will only ever read the first we receive. We will wait to receive this before populating movement groups or processing any other incoming messages since this information is required for all of our signal optimization logic. If it fails the service should fail or stop.

File: signal_opt_service/src/signal_opt_service.cpp (hunk 14 above)
Comment (paulbourelly999) [reply signal_opt_service/src/signal_opt_service.cpp:99]: We are not consuming from all kafka topics the same way. For tsc_state configuration we only need a single update but for spat and vsi we continuously update.

File: signal_opt_service/src/signal_opt_service.cpp (hunk 15 above)
Comment (paulbourelly999) [reply signal_opt_service/src/signal_opt_service.cpp:182]: Will correct

File: signal_opt_service/test/test_signal_opt_messages_worker.cpp (hunk 16 above)
Comment (paulbourelly999) [reply signal_opt_service/test/test_signal_opt_messages_worker.cpp:20]: Any debug log statements will never print. I should remove this log statement though anyway.

File: streets_utils/streets_signal_phase_and_timing/src/models/spat.cpp [hunk 17]
```diff
@@ -2,11 +2,13 @@
 
 namespace signal_phase_and_timing{
 
-    std::string spat::toJson() const {
+    std::string spat::toJson() {
```
Comment (dan-du-car) [inline streets_utils/streets_signal_phase_and_timing/src/models/spat.cpp:5]: Not sure why const is removed

**Review by Saeid-Solei — APPROVED** [review]
